pkg/sql/sqlutil/internal_executor.go

```diff
@@ -191,16 +212,74 @@ type InternalRows interface {
 	Types() colinfo.ResultColumns
 }
 
-// SessionBoundInternalExecutorFactory is a function that produces a "session
+// InternalExecutorProto stores info needed to initialize an
+// internal executor.
+// The only difference between InternalExecutorProto and InternalExecutorFactory
+// is that the proto also includes the synthetic descriptor.
+type InternalExecutorProto struct {
+	IeFactory InternalExecutorFactory
+	// SyntheticDescs stores the synthetic descriptors to be injected into
+	// each query/statement's descs.Collection upon initialization.
+	//
+	// Warning: Not safe for concurrent use from multiple goroutines.
+	SyntheticDescs []catalog.Descriptor
+}
+
+// SetSyntheticDescs is to set the synthetic descriptor stored in the internal
+// executor proto.
+func (p *InternalExecutorProto) SetSyntheticDescs(d []catalog.Descriptor) {
+	p.SyntheticDescs = d
+}
+
+// InternalExecutorFactory is a function that produces a "session
 // bound" internal executor.
-type SessionBoundInternalExecutorFactory func(
+type InternalExecutorFactory func(
 	context.Context, *sessiondata.SessionData,
 ) InternalExecutor
 
+// WithTxn is to run SQL statements under the same txn.
+// All executions within this scope should share the same descriptor
+// collection.
+// TODO (janexing): verify if this is the desired behavior.
```

Comment:
> `WithTxn` actually creating a desc collection and setting the txn here

Yes this is the intent -- we want to make sure that if multiple SQL statements are run within the same txn, they should share the descriptor collection. It means that we always pass the same collection pointer to init the child conn executor under the same ie. In other words, the collection now is more closely binded with an internal executor, instead of a child conn executor.